File: applications/rtkdrawgeometricphantom/rtkdrawgeometricphantom.cxx
/*=========================================================================
 *
 *  Copyright RTK Consortium
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *         https://www.apache.org/licenses/LICENSE-2.0.txt
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *=========================================================================*/

#include "rtkdrawgeometricphantom_ggo.h"
#include "rtkGgoFunctions.h"
#include "rtkDrawGeometricPhantomImageFilter.h"
#include "itkImageFileWriter.h"
#include "rtkAdditiveGaussianNoiseImageFilter.h"


int
main(int argc, char * argv[])
{
  GGO(rtkdrawgeometricphantom, args_info);

  using OutputPixelType = float;
  constexpr unsigned int Dimension = 3;
  using OutputImageType = itk::Image<OutputPixelType, Dimension>;

  // Empty volume image
  using ConstantImageSourceType = rtk::ConstantImageSource<OutputImageType>;
  auto constantImageSource = ConstantImageSourceType::New();
  rtk::SetConstantImageSourceFromGgo<ConstantImageSourceType, args_info_rtkdrawgeometricphantom>(constantImageSource,
                                                                                                 args_info);

  using DQType = rtk::DrawGeometricPhantomImageFilter<OutputImageType, OutputImageType>;

  // Offset, scale, rotation
  DQType::VectorType offset(0.);
  if (args_info.offset_given)
  {
    if (args_info.offset_given > 3)
    {
      std::cerr << "--offset needs up to 3 values" << std::endl;
      exit(EXIT_FAILURE);
    }
    offset[0] = args_info.offset_arg[0];
    offset[1] = args_info.offset_arg[1];
    offset[2] = args_info.offset_arg[2];
  }
  DQType::VectorType scale;
  scale.Fill(args_info.phantomscale_arg[0]);
  if (args_info.phantomscale_given)
  {
    if (args_info.phantomscale_given > 3)
    {
      std::cerr << "--phantomscale needs up to 3 values" << std::endl;
      exit(EXIT_FAILURE);
    }
    for (unsigned int i = 0; i < std::min(args_info.phantomscale_given, Dimension); i++)
      scale[i] = args_info.phantomscale_arg[i];
  }
  DQType::RotationMatrixType rot;
  rot.SetIdentity();
  if (args_info.rotation_given)
  {
    if (args_info.rotation_given != 9)
    {
      std::cerr << "--rotation needs exactly 9 values" << std::endl;
      exit(EXIT_FAILURE);
    }
    for (unsigned int i = 0; i < Dimension; i++)
      for (unsigned int j = 0; j < Dimension; j++)
        rot[i][j] = args_info.rotation_arg[i * Dimension + j];
  }

  // Reference
  if (args_info.verbose_flag)
    std::cout << "Creating reference... " << std::flush;
  auto dq = DQType::New();
  dq->SetInput(constantImageSource->GetOutput());
  dq->SetPhantomScale(scale);
  dq->SetOriginOffset(offset);
  dq->SetRotationMatrix(rot);
  dq->SetConfigFile(args_info.phantomfile_arg);
  TRY_AND_EXIT_ON_ITK_EXCEPTION(dq->Update())

  // Add noise
  OutputImageType::Pointer output = dq->GetOutput();
  if (args_info.noise_given)
  {
    using NIFType = rtk::AdditiveGaussianNoiseImageFilter<OutputImageType>;
    auto noisy = NIFType::New();
    noisy->SetInput(output);
    noisy->SetMean(0.0);
    noisy->SetStandardDeviation(args_info.noise_arg);
    TRY_AND_EXIT_ON_ITK_EXCEPTION(noisy->Update())
    output = noisy->GetOutput();
  }

  // Write
  if (args_info.verbose_flag)
    std::cout << "Writing reference... " << std::flush;
  TRY_AND_EXIT_ON_ITK_EXCEPTION(itk::WriteImage(output, args_info.output_arg))

  return EXIT_SUCCESS;
}

File: applications/rtkdrawgeometricphantom/rtkdrawgeometricphantom.py
#!/usr/bin/env python
import argparse
import sys
import itk
from itk import RTK as rtk
import numpy as np


def build_parser():
    # Argument parsing
    parser = rtk.rtk_argument_parser("Computes a 3D voxelized phantom from a phantom description file.")

    # General options
    parser.add_argument(
        "--verbose", "-v", help="Verbose execution", action="store_true"
    )
    parser.add_argument(
        "--output", "-o", help="Output projections file name", type=str, required=True
    )
    parser.add_argument(
        "--phantomfile",
        help="Parameters of the phantom reference",
        type=str,
        required=True,
    )
    parser.add_argument("--noise", help="Gaussian noise parameter (SD)", type=float)
    parser.add_argument(
        "--phantomscale",
        help="Scaling factor for the phantom dimensions",
        type=rtk.comma_separated_args(float),
        default=[1.0],
    )
    parser.add_argument(
        "--offset",
        help="3D spatial offset of the phantom center",
        type=rtk.comma_separated_args(float),
    )
    parser.add_argument(
        "--rotation",
        help="Rotation matrix for the phantom",
        type=rtk.comma_separated_args(float),
    )

    rtk.add_rtk3Doutputimage_group(parser)

    # Parse the command line arguments
    return parser


def process(args_info: argparse.Namespace):
    # Define output pixel type and dimension
    OutputPixelType = itk.F
    Dimension = 3
    OutputImageType = itk.Image[OutputPixelType, Dimension]

    # Empty volume image
    constant_image_source = rtk.ConstantImageSource[OutputImageType].New()
    rtk.SetConstantImageSourceFromArgParse(constant_image_source, args_info)

    # Offset, scale, rotation
    offset = [0.0] * 3
    if args_info.offset:
        if len(args_info.offset) > 3:
            print("--offset needs up to 3 values", file=sys.stderr)
            sys.exit(1)
        offset = args_info.offset

    scale = [args_info.phantomscale[0]] * Dimension
    if args_info.phantomscale:
        if len(args_info.phantomscale) > 3:
            print("--phantomscale needs up to 3 values", file=sys.stderr)
            sys.exit(1)
        for i in range(len(args_info.phantomscale)):
            scale[i] = args_info.phantomscale[i]

    rot = itk.Matrix[itk.D, Dimension, Dimension]()
    rot.SetIdentity()
    if args_info.rotation:
        if len(args_info.rotation) != 9:
            print("--rotation needs exactly 9 values", file=sys.stderr)
            sys.exit(1)
        itk.matrix_from_array(np.array(args_info.rotation).reshape(3, 3))

    # Reference
    if args_info.verbose:
        print("Creating reference... ", flush=True)

    # DrawGeometricPhantomImageFilter
    dq = rtk.DrawGeometricPhantomImageFilter[OutputImageType, OutputImageType].New()
    dq.SetInput(constant_image_source.GetOutput())
    dq.SetPhantomScale(scale)
    dq.SetOriginOffset(offset)
    dq.SetRotationMatrix(rot)
    dq.SetConfigFile(args_info.phantomfile)

    dq.Update()

    # Add noise
    output = dq.GetOutput()
    if args_info.noise:
        noisy = rtk.AdditiveGaussianNoiseImageFilter[OutputImageType].New()
        noisy.SetInput(output)
        noisy.SetMean(0.0)
        noisy.SetStandardDeviation(args_info.noise)

        noisy.Update()

        output = noisy.GetOutput()

    # Write
    if args_info.verbose:
        print("Writing reference... ", flush=True)

    itk.imwrite(output, args_info.output)


def main(argv=None):
    parser = build_parser()
    args_info = parser.parse_args(argv)
    process(args_info)


if __name__ == "__main__":
    main()
